---
title: "Gastrulation scNMT-seq: correlation per gene (across cells) between RNA expression and DNA methylation"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(weights))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(data.table))
```

<!-- Define functions -->
```{r echo=FALSE, include=FALSE}
source("/Users/ricard/gastrulation/metrna/cor/utils.R")
```

<!-- Define I/O and options -->
```{r echo=TRUE, include=FALSE}

## I/O ##
io   <- list()
io$basedir   <- "/Users/ricard/data/gastrulation"
io$sample.metadata <- paste0(io$basedir,"/sample_metadata.txt")
io$annos_dir  <- paste0(io$basedir, "/features/genomic_contexts")
io$rna_file   <- paste0(io$basedir, "/rna/SingleCellExperiment.rds")
io$met_dir   <- paste0(io$basedir, "/met/feature_level")
io$outdir <- paste0(io$basedir,"/metrna/cor")
io$gene_metadata <- "/Users/ricard/data/ensembl/mouse/v87/BioMart/mRNA/Mmusculus_genes_BioMart.87.txt"

## Options ##
opts <- list()

# Filtering parameters
opts$min.CpGs <- 1      # Minimum number of CpGs per feature
opts$min.cells <- 50    # Minimum number of observed cells per feature with at least opts$min.CpGs
opts$max.genes <- 5000  # Top N most variable genes and features

# Multiple testing correction options
opts$threshold_fdr  <- 0.1

# Correlation type options
opts$method <- "pearson"      # correlation type (see ?cor)
opts$weight <- FALSE          # weighted correlation (see ?wtd.cor) 

# Permutation test options
opts$permutation <- TRUE   # do permutation test?
opts$n_perms <- 10          # Number of random permutations

# Define genomic contexts
opts$annos <- c(
  "prom_2000_2000"="Promoters"
)

# Define stage and lineages
opts$stage_lineage <- c(

  # E4.5
  "E4.5_Epiblast",

  # E5.5
  "E5.5_Epiblast",
  
  # E6.5
  "E6.5_Epiblast",
  "E6.5_Primitive_Streak",
  
  # E7.5
  "E7.5_Epiblast",
  "E7.5_Ectoderm",
  "E7.5_Primitive_Streak",
  "E7.5_Endoderm",
  "E7.5_Mesoderm"
)

# Define which cells to use
tmp <- fread(io$sample.metadata) %>% 
  .[,stage_lineage:=paste(stage,lineage10x_2,sep="_")] %>%
  .[pass_metQC==T & pass_rnaQC==T & stage_lineage%in%opts$stage_lineage] 
opts$rna.cells <- tmp$id_rna
opts$met.cells <- tmp$id_met
```

<!-- Load sample metadata -->
```{r}
sample_metadata <- fread(io$sample.metadata) %>% 
  .[,c("sample","id_rna","id_met","stage","lineage10x_2")] %>%
  .[,stage_lineage:=paste(stage,lineage10x_2,sep="_")] %>%
  .[sample%in%tmp$sample]
```

<!-- Load RNA data -->
```{r load_rna, echo=FALSE, include=FALSE}

# Load SingleCellExperiment object
sce <- readRDS(file=io$rna_file)[,opts$rna.cells]

# Create data.table
rna_dt <- exprs(sce) %>% t %>% as.data.table(keep.rownames="id_rna") %>% 
  melt(id.vars="id_rna", value.name="expr", variable.name="id") %>%
  merge(sample_metadata[,c("sample","id_rna")]) %>%
  merge(rowData(sce) %>% as.data.frame(row.names=rownames(sce)) %>% tibble::rownames_to_column("ens_id") %>%
          .[,c("symbol","ens_id")] %>% as.data.table %>% setnames(c("ens_id","symbol"),c("id","gene")),
        by=c("id")
        )
rna_dt[,c("id","gene"):=list(as.factor(id),as.factor(gene))]
```


<!-- Filter RNA data -->
```{r}
# Remove genes with little variability
rna_dt <- rna_dt[,var:=var(expr), by="gene"] %>% .[var>0.5] %>% .[,var:=NULL]
```

<!-- Load methylation data -->
```{r load_met, echo=FALSE, include=FALSE}
met_dt <- lapply(names(opts$annos), function(n) {
  fread(sprintf("%s/%s.tsv.gz",io$met_dir,n)) %>% .[V1%in%opts$met.cells]
}) %>% rbindlist %>% setnames(c("sample","id","anno","Nmet","N","rate"))
```

<!-- Filter methylation data -->
```{r load_met, echo=FALSE, include=FALSE}
# Filter by mininum number of CpGs per feature
met_dt <- met_dt[N>=opts$min.CpGs]

# Filter by coverage
met_dt <- met_dt %>% .[,N:=.N, by=c("id","anno")] %>% .[N>=opts$min.cells] %>% .[,N:=NULL]
```

<!-- Load gene metadata  -->
```{r load_genomiccontexts}
gene_metadata <- fread(io$gene_metadata) %>% 
  setnames(c("ens_id","symbol"),c("id","gene")) %>% 
  .[,chr:=stringr::str_replace_all(chr,"chr","")] 
```

<!-- Merge DNA methylation and RNA expression data -->
```{r}
met_dt <- met_dt %>% merge(gene_metadata[,c("id","gene")], by="id")

metrna_dt <- merge(
  met_dt, 
  rna_dt[,c("sample","gene","expr")], 
  by=c("sample","gene")
)
```

<!-- Filter data -->
```{r filter_across_samples, echo=FALSE}
# Filter by variance
metrna_dt <- metrna_dt[,var:=var(rate), by=c("id","anno","gene")] %>% .[var>0] %>% .[,var:=NULL]
metrna_dt <- metrna_dt[,var:=var(expr), by=c("id","anno","gene")] %>% .[var>0] %>% .[,var:=NULL]

# Extract top N most variable genes
keep_hv_genes <- metrna_dt %>% split(.$anno) %>% 
  map(~ .[,.(var=var(expr)), by=c("gene","id")] %>% setorder(-var) %>% head(n=opts$max.genes) %>% .[,gene_id:=paste(gene,id,sep="_")] %>% .$gene_id %>% as.character())
metrna_dt <- metrna_dt %>% .[,gene_id:=paste(gene,id,sep="_")] %>% split(.$anno) %>%
  map2(.,names(.), function(x,y) x[gene_id %in% keep_hv_genes[[y]]]) %>% rbindlist %>% .[,gene_id:=NULL]
```

<!-- Compute correlations and do multiple testing correction -->

```{r}
if (opts$weight) {
  cor <- metrna_dt[, wtd.cor(rate, expr, N)[,c("correlation","t.value","p.value")], by = c("id","gene","anno")]
} else {
  cor <- metrna_dt[, .(V1 = unlist(cor.test(rate, expr, method = opts$method)[c("estimate", "statistic", "p.value")])), by = c("id","gene","anno")]
}

# Compute adjusted p-values (both FDR and Bonferroni)
cor <- cor %>% .[, para := rep(c("r","t","p"), .N/3)] %>% data.table::dcast(id+gene+anno ~ para, value.var = "V1") %>%
      .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method="fdr"), p.adjust(p, method="bonferroni")), by="anno"] %>%
      .[, c("log_padj_fdr","log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
      .[, sig := padj_fdr <= opts$threshold_fdr] %>%  setorder(padj_fdr)
```

```{r}
cor[,sum(sig),by="anno"]
```

<!-- Save results -->
```{r}
# fwrite(cor, paste0(io$outdir,"/metrna_cor_promoters.txt"), quote=F, sep="\t")
# cor <- fread(paste0(io$outdir,"/metrna_cor_promoters.txt"))
```

<!-- Run permutation test for the correlation across samples -->
```{r perm_cor_rate, echo=FALSE, include=FALSE}
if (opts$permutation) {
  pp_vals <- vector(mode = "numeric", length = length(cor$p))
  # For each iteration...
  for (k in 1:opts$n_perms){
    print(k)
    # Copy original methylation data
    metrna_dt_perm <- copy(metrna_dt)
    # Permute gene expression levels
    metrna_dt_perm <- metrna_dt_perm[, expr:=sample(expr), by = c("id","gene","anno")]
    # Permute methylation levels
    metrna_dt_perm <- metrna_dt_perm[, rate:=sample(rate), by = c("id","gene","anno")]
    # Compute correlation across samples
    if (opts$weight) {
      # cor_perm <- metrna_dt_perm[, .(p = wtd.cor(rate, expr, weight)[, c("p.value")]), by = c("id", "gene", "anno")]
    } else {
      cor_perm <- metrna_dt_perm[, .(p = cor.test(rate, expr, method=opts$method)[["p.value"]]), by = c("id","gene","anno")]
    }
    
    # For each annotation sort by p-value and store the permuted p-values
    cor_perm <- cor_perm %>% split(.$anno) %>% map(~ .[,.(anno = anno, p = sort(p))]) %>% rbindlist
    pp_vals <- pp_vals + cor_perm$p
  }
  # Compute the average p-values
  pp_vals <- pp_vals / opts$n_perms
  # Store them in a data.table for later analysis
  cor_perm <- cor_perm[, p := pp_vals]
}
```

<!-- Volcano plot of p-values against Pearson's r and QQ-plot of p-values -->
```{r, echo=FALSE}
if (opts$permutation) {
  pp <- qq <- list()
  for (n in unique(cor$anno)) {
    
    # Compute positive and negative hits in terms of correlation
    negative_hits <- cor[anno==n & sig==TRUE & r<0,id]
    positive_hits <- cor[anno==n & sig==TRUE & r>0,id]
    all <- nrow(cor[anno == n,])
    
    # Generate volcano plot  
    tmp <- cor
    tmp[log_padj_fdr>7,log_padj_fdr:=7]
    pp[[n]] <- gg_volcano_plot(tmp[anno == n,], title = "", label=10)
    
    # Generate permutation plot
    qq[[n]] <- gg_qqplot(cor[anno == n, ], cor_perm[anno == n]$p, title = "")
    
    
    # Plot volcano + qqplot
    # pdf(file=paste0(io$outdir,"/pdf/volcano_qq_", n, ".pdf"), width = 13, height = 5.5, useDingbats = FALSE)
    grid.arrange(pp[[n]], qq[[n]], ncol=2, top = textGrob(n, gp=gpar(fontsize=29, fontface = "bold")), newpage = TRUE)
    # dev.off()
  }
}
```
