---
title: "Gastrulation:"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(ggplot2)
library(scater)


theme_pb <- function() {
  theme(
    axis.text.x = element_text(size=rel(1.4), color="black", angle=50, vjust=1, hjust=1),
    axis.text.y = element_text(size=rel(1.2), color="black"),
    axis.title.y = element_text(size=rel(1.4), color="black"),
    strip.background = element_rect(fill="#F37A71"),
    strip.text = element_text(size=rel(1.8), color="black"),
    legend.position="top",
    legend.title = element_blank()
  )
}
```

<!-- Define I/O and options -->
```{r define_opts, echo=FALSE}
source("/Users/ricard/gastrulation/metacc/boxplots_enhancers/load_settings.R")
```

<!-- Load sample metadata -->
```{r load_metadata}
sample_metadata <- fread(io$sample.metadata) %>%
  .[,c("sample","id_acc","id_met","id_rna","stage","lineage10x_2","plate")] %>%
  .[,lineage10x_2:=stringr::str_replace_all(lineage10x_2,"_"," ")] %>% 
  .[,stage_lineage:=paste(stage,lineage10x_2,sep=" ")] %>%
  .[id_met%in%opts$met_cells | id_acc%in%opts$acc_cells]
```

<!-- Load methylation data -->
```{r load_data, echo=FALSE, include=FALSE}
met_dt <- lapply(names(opts$met.annos), function(n) {
  data <- fread(sprintf("%s/%s.tsv.gz",io$met.dir,n)) %>% .[V1%in%opts$met_cells]
}) %>% rbindlist %>% setnames(c("id_met","id","anno","Nmet","Ntotal","rate"))
```

<!-- Load accessibility data -->
```{r load_data, echo=FALSE, include=FALSE}
acc_dt <- lapply(names(opts$acc.annos), function(n) {
  data <- fread(sprintf("%s/%s.tsv.gz",io$acc.dir,n)) %>% .[V1%in%opts$acc_cells]
}) %>% rbindlist %>% setnames(c("id_acc","id","anno","Nmet","Ntotal","rate"))
```

<!-- Merge data with metadata -->
```{r merge}
acc_dt <- merge(acc_dt, sample_metadata, by="id_acc") %>% droplevels()
met_dt <- merge(met_dt, sample_metadata, by="id_met") %>% droplevels()
```

```{r}
met_dt <- met_dt %>% .[,anno:=stringr::str_replace_all(anno,opts$met.annos)]
acc_dt <- acc_dt %>% .[,anno:=stringr::str_replace_all(anno,opts$acc.annos)]
```

<!-- Load results from the differential analysis -->
```{r}
source("/Users/ricard/gastrulation/metaccrna/differential/load_data.R")

diff.met <- diff.met %>%
  .[,anno:=stringr::str_replace_all(anno,opts$met.annos)] %>%
  .[,c("id","anno","diff","sig","lineage")]

diff.acc <- diff.acc %>%
  .[,anno:=stringr::str_replace_all(anno,opts$acc.annos)] %>%
  .[,c("id","anno","diff","sig","lineage")]
```

<!-- Subset lineage-defining sites -->
Lineage-defining elements are defined as ChIP-seq peaks that show differential activity during germ layer commitment

Methylation
```{r}
met_dt <- met_dt %>% split(.$anno) %>%
  map2(.,names(.), function(x,y) x[id%in%diff.met[sig==T & anno==y,id]]) %>%
  rbindlist %>% droplevels()
```

Accessibility
```{r}
acc_dt <- acc_dt %>% split(.$anno) %>%
  map2(.,names(.), function(x,y) x[id%in%diff.acc[sig==T & anno==y,id]]) %>%
  rbindlist %>% droplevels()
```

```{r}
# fwrite(diff.met[sig==T], "/Users/ricard/data/gastrulation/met/differential/diffmet.txt", sep="\t")
# fwrite(diff.acc[sig==T], "/Users/ricard/data/gastrulation/met/differential/diffacc.txt", sep="\t")
```

<!-- Filter by minimum number of measurements per cell -->
```{r}
opts$min.acc.observations <- 50
acc_dt <- acc_dt[,N:=.N, by=c("sample","stage_lineage","anno")] %>% .[N>=opts$min.acc.observations] %>% .[,N:=NULL]

opts$min.met.observations <- 10
met_dt <- met_dt[,N:=.N, by=c("sample","stage_lineage","anno")] %>% .[N>=opts$min.met.observations] %>% .[,N:=NULL]
```

<!-- Regress out differences in global methylation or accessibility rate -->
```{r}
# foo <- fread(io$met.stats) %>% .[,c("id_met","mean")]
# met_dt <- met_dt %>% merge(foo, by="id_met") %>%
#   .[,new_rate:=mean(rate) + lm(formula=rate~mean)[["residuals"]], by=c("id","anno")]
# 
# foo <- fread(io$acc.stats) %>% .[,c("id_acc","mean")]
# acc_dt <- acc_dt %>% merge(foo, by="id_acc") %>%
#   .[,new_rate:=mean(rate) + lm(formula=rate~mean)[["residuals"]], by=c("id","anno")]
```

<!-- Boxplots with accessibility rate per genomic context and stage_lineage -->
```{r}
foo.acc <- acc_dt %>%
  .[,.(rate=100*(sum(Nmet)/sum(Ntotal)), Nmet=sum(Nmet), N=sum(Ntotal)),by=c("sample","stage_lineage","anno")] %>%
  # .[,.(rate=mean(new_rate)),by=c("id_acc","stage_lineage","anno")] %>%
  .[,stage_lineage:=factor(stage_lineage, levels=names(opts$colors))]

p <- ggplot(foo.acc, aes(x=stage_lineage, y=rate)) +
  geom_boxplot(aes(fill=stage_lineage), outlier.shape=NA, coef=1) +
  scale_fill_manual(values=opts$colors) +
  labs(x="", y="Accessibility (%)") +
  facet_wrap(~anno, nrow=1, scales="fixed") +
  coord_cartesian(ylim=c(21,55)) +
  theme_bw() +
  guides(color=F, fill=F) +
  theme_pb() +
  theme(strip.background = element_rect(fill="#00BFC4"))
print(p)

# pdf(paste0(io$outdir,"/boxplots_acc2.pdf"), useDingbats = F, onefile = F, width=14, height=6)
# print(p)
# dev.off()
```

<!-- Boxplots with methylation rate per genomic context and stage_lineage -->
```{r}
foo.met <- met_dt %>%
  .[,.(rate=100*(sum(Nmet)/sum(Ntotal)), Nmet=sum(Nmet), N=sum(Ntotal)),by=c("sample","stage_lineage","anno")] %>%
  # .[,.(rate=mean(rate)),by=c("id_acc","stage_lineage","anno")] %>%
  .[,stage_lineage:=factor(stage_lineage, levels=names(opts$colors))]

p <- ggplot(foo.met, aes(x=stage_lineage, y=rate)) +
  geom_boxplot(aes(fill=stage_lineage), outlier.shape=NA, coef=1) +
  scale_fill_manual(values=opts$colors) +
  labs(x="", y="Methylation (%)") +
  facet_wrap(~anno, nrow=1, scales="fixed") +
  coord_cartesian(ylim=c(8,93)) +
  theme_bw() +
  guides(color=F, fill=F) +
  theme_pb() +
  theme(strip.background = element_rect(fill="#F37A71"))
print(p)

# pdf(paste0(io$outdir,"/boxplots_met2.pdf"), useDingbats = F, onefile = F, width=14, height=6)
# print(p)
# dev.off()
```
